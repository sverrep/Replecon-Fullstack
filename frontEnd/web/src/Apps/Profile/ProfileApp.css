.content{

}

ul li {
    box-shadow: 0 4px 8px 0 rgba(0,0,0,0.2);
    transition: 0.3s;
    padding: 3px 8px;
    border-radius: 5px;
    margin-top: 2px;
    
  }

  .green{
    background-color:rgb(49, 211, 49);
  }

  .red{
    background-color:rgb(253, 63, 63);
  }

  .grey{
    background-color:rgb(117, 117, 117);
  }

.transactions{
    width: 50%;
    float: left;
}

.boughtItems{
    width: 50%;
    float: right;
}