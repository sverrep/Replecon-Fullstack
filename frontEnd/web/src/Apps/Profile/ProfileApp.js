import React from 'react';
import { withRouter, Redirect } from "react-router-dom";
import axios from 'axios';
import getIP from '../../settings.js';
import './ProfileApp.css';
import navbar from '../../Components/navbar/Student NavBar/Navbar.js';
import Button from 'react-bootstrap/Button';
import Container from 'react-bootstrap/Container'
import Row from 'react-bootstrap/Row'
import ListGroup from 'react-bootstrap/ListGroup'
import Modal from 'react-bootstrap/Modal'
import Form from 'react-bootstrap/Form'
import FormControl from 'react-bootstrap/FormControl'
import FloatingLabel from 'react-bootstrap/FloatingLabel'

class Profile extends React.Component {
    constructor(props) {
        super(props);
        this.state = { 
            email: '', 
            password: '', 
            first_name: '', 
            last_name: '', 
            user_id: '', 
            teacher_id: '', 
            class_code: '', 
            balance: '',
            new_class_name: '',
            transactions: [],  
            bought_items: [], 
            role: this.props.location.state.role, 
            classes: [], 
            redirect_login: false, 
            redirect_class: false, 
            selected_class: [],
            showCreateClass: false,
            token: this.props.location.state.token,
            showInven:false,
        };
        this.handleLogOut = this.handleLogOut.bind(this)
        this.handleChange = this.handleChange.bind(this)
        this.openInven = this.openInven.bind(this)
        this.closeInven = this.closeInven.bind(this)
    }

    componentDidMount(){
        axios.defaults.headers.common.Authorization = `Token ${this.state.token}`;
        axios.get(getIP()+'/students/current/')
        .then(response => {
            if(this.state.role === "Student")
            {
                this.setState({first_name: response.data.first_name, user_id: response.data.id, email: response.data.username})
                this.getStudentBalance()
                this.getStudentTransactions()
                this.getBoughtItems()
            }
            else if(this.state.role === "Teacher")
            {
                this.getTeacherClasses()
                this.setState({first_name: response.data.user.first_name, last_name: response.data.teacher.last_name, user_id: response.data.user.id, email: response.data.user.username, teacher_id: response.data.teacher.id})
            }
        })
        .catch(error => console.log(error))
    }

    handleChange(e) {
        const field = e.target.id
        this.setState({ [field] : e.target.value });
    }

    getTeacherClasses(){
        axios.get(getIP()+'/classrooms/')
        .then(response => {
            this.findTeacherClassrooms(response.data)
        })
        .catch(error => console.log(error))
      }
    
      findTeacherClassrooms(classrooms){
        for (let i = 0; i <= Object.keys(classrooms).length-1; i++)
        {
          if (classrooms[i].teacher_id === this.state.teacher_id)
          {
            this.setState({ classes: [...this.state.classes, classrooms[i]] });
          }
        }
      }

    handleLogOut() {
        axios.get(getIP()+'/auth/logout/')
          .then(response => {
            axios.defaults.headers.common.Authorization = null
            this.setState({ redirect_login : true})
          })
          .catch(error =>  console.log(error));
      }
      
    handleClassRedirect(item){
        this.setState({class_code: item.class_code, redirect_class: true, selected_class: item})
      }

    getStudentBalance() {
     axios.get(getIP()+'/students/balance/')
    .then(response => {
        const balance = response.data
        this.setState({balance})
    })
    .catch(error =>  console.log(error));
    }

    getStudentTransactions(){
        axios.get(getIP()+'/transactions/getAllStudentTransactions/')
        .then(response =>{
          this.setState({transactions: response.data})
        })
    .catch(error => console.log(error))
    }

    getBoughtItems(){
        axios.get(getIP()+'/items/boughtitems/')
        .then(response => {
            this.setState({bought_items: response.data})
        })
        
        .catch(error => console.log(error))
    }

    createClass(){
        axios.get(getIP()+'/classrooms/')
        .then(response => {
          this.setState({new_class_code: this.makeClassCode(response.data)}, () => {
            const payload = {class_name: this.state.new_class_name, teacher_id: this.state.teacher_id, class_code: this.state.new_class_code}
            axios.post(getIP()+'/classrooms/', payload)
            .then(response => {
              this.setState({showCreateClass: false}, () => {
                this.setState({ classes: [...this.state.classes, response.data] });
              })
            })
            .catch(error => console.log(error))
         })
        })
       .catch(error => console.log(error))
     }

     
    makeClassCode(classrooms) {
        var result = []
        var characters = 'ABCDEFGHIJKLMNPQRSTUVWXYZ123456789'
        var code_length = 6
        var alreadyExists = false
        
        for ( let i = 0; i < code_length;i++){
            result.push(characters.charAt(Math.floor(Math.random() * characters.length)));
        }
        for (let k = 0; k<=Object.keys(classrooms).length -1;k++)
        {
        if(classrooms[k].class_code === result)
        {
            alreadyExists = true
        }
        }

        if (alreadyExists === false){
            return result.join('')
        }
        else{
            this.makeClassCode(classrooms)
        }
    }


    renderCard(item){
        if(item.symbol === '+'){
            var transCardplus = 'From ' + item.name + " " +  item.symbol + item.amount +"$"
            return(transCardplus)
        }
        else{
            var transCardminus = 'To ' + item.name + " " +  item.symbol + item.amount +"$"
            return(transCardminus)
        }
        }
        
        
        
    

    renderItemCard(item){
        return item.item_name
    }

    createClassModal() {
        return (
          <Modal
            show={this.state.showCreateClass}
            onHide={() => this.setState({showCreateClass: false})}
            size="lg"
            centered
          >
            <Modal.Header closeButton>
              <Modal.Title>
                Create A Class
              </Modal.Title>
            </Modal.Header>
            <Modal.Body>
              <FloatingLabel label="Class Name">
                <FormControl id='new_class_name' placeholder="Class Name" onChange={this.handleChange}/>
              </FloatingLabel>
            </Modal.Body>
            <Modal.Footer>
              <Button className="modal-btn" onClick={() => this.createClass()}>Create Class</Button>
            </Modal.Footer>
          </Modal>
        );
      }
    
    openInven(){
        this.setState({showInven:true})
    }
    closeInven(){
        this.setState({showInven:false})
    }
    renderInventoryModal(){
        return(
            <Modal
            show={this.state.showInven}
            
            keyboard={false}
            >
                <Modal.Header>
                    <Modal.Title>My Inventory</Modal.Title>
                    <Button variant="secondary" onClick={() => this.closeInven()}>Close</Button>
                </Modal.Header>

                <Modal.Body>
                    <Form>
                    <div className='boughtItems'>
                            <ul>
                                {this.state.bought_items.map(item => {
                                    return <li key={item.id} className='grey cardb'>{this.renderItemCard(item)}</li>;
                                })}
                            </ul>
                        </div>
                    </Form>
                </Modal.Body>
            </Modal>
        )
    }
    render() {
        if(this.state.redirect_login){
            return(
                <Redirect to={{pathname: '/Login'}}></Redirect>
            )
        }
        else if(this.state.redirect_class){
            return(
                <Redirect to={{pathname: `/Class/${this.state.class_code}/students`, state: {class: this.state.selected_class, teacher_id: this.state.teacher_id, token: this.state.token}}}></Redirect>
            );
        }
        else if (this.state.role === "Student"){
            return (
                <div className='wrapper'>
                    {navbar(this.state.token)}
                    <div className='profile-content'>
                    <div className='profile-title'>
                        <div className='profile-name'><h3>{this.state.first_name}</h3></div>
                        
                        <div className='profile-balance'>
                            <Button variant='primary' className='profile-button' onClick={this.handleLogOut}>
                            Log Out
                            </Button>

                            <Button variant='primary' className='inventory-button' onClick={this.openInven}>
                            Inventory
                            </Button>
                            
                        </div>
                        
                    {this.renderInventoryModal()}
                    </div>
                    <h3>${this.state.balance}</h3>
                    <div className='content'>
                        <div>
                            <ul>
                                {this.state.transactions.map(item => {
                                    if(item.symbol === '+'){
                                        return <li key={item.id} className='green cardb'>{this.renderCard(item)}</li>;
                                    }
                                    else{
                                        return <li key={item.id} className='red cardb'>{this.renderCard(item)}</li>;
                                    }
                                })}
                            </ul>   
                        </div>

                        
                    </div>
                    </div>
                </div>
            );
        }
        else if (this.state.role === "Teacher"){
            return (
            <Container className="teacher-container">
                <Row>
                    <h3 className="teacher-h3">Welcome Back {this.state.last_name}</h3>
                    <h4>Your Classes</h4>
                        {this.state.classes.map((item,i) => <ListGroup.Item key={i} className="class-list" action onClick={() => this.handleClassRedirect(item)}>{item.class_name} {item.class_code}</ListGroup.Item>)}
                    <div className="teacher-btns-row">
                        <Button variant="primary" className="teacher-btn-create" onClick={() => this.setState({showCreateClass: true})}>
                            Create New Class
                        </Button>
                    </div>
                    <div className="teacher-btns-row">
                        <Button variant="outline-secondary" className="teacher-btn-logout" onClick={this.handleLogOut}>
                            Log Out
                        </Button>
                    </div>
                </Row>
                {this.createClassModal()}
            </Container>
            );
        }
            
        
    }
}

export default withRouter(Profile)