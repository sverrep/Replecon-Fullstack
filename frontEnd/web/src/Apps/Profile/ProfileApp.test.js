import React from 'react';
import ProfileApp from './ProfileApp.js'
import axios from 'axios'

jest.mock('axios');
let wrapper
let wrapperTeacher
const props = {
    location:{
        state:{
            role: "Student",
            token: 'Token'
        }
    }  
}

const propsTeacher = {
    location:{
        state:{
            role: "Teacher",
            token: 'Token'
        }
    }  
}


describe('Student Profile Tests', ()=>{
    
    beforeEach(() => {
        wrapper = shallow(<ProfileApp.WrappedComponent {...props}/>, { disableLifecycleMethods: true })
        wrapperTeacher = shallow(<ProfileApp.WrappedComponent {...propsTeacher}/>, { disableLifecycleMethods: true })
    });

    describe('Component did mount', ()=>{
        it('should call component did mount properly as a student', async()=>{
            const instance = wrapper.instance()
            
            const user = mockResponseData({id: '1', first_name: 'name', username: 'email@gmail.com'})
            axios.get.mockImplementationOnce(() =>  Promise.resolve(user))

            instance.getStudentBalance = jest.fn()
            instance.getStudentBalance.mockReturnValueOnce()

            instance.getStudentTransactions = jest.fn()
            instance.getStudentTransactions.mockReturnValueOnce()

            instance.getBoughtItems = jest.fn()
            instance.getBoughtItems.mockReturnValueOnce()

            await instance.componentDidMount()

            expect(instance.state.first_name).toBe('name')
            expect(instance.getStudentBalance).toHaveBeenCalledTimes(1)
            expect(instance.getStudentTransactions).toHaveBeenCalledTimes(1)
            expect(instance.getBoughtItems).toHaveBeenCalledTimes(1)
        })

        it('should call component did mount properly as a teacher', async()=>{
            const instance = wrapperTeacher.instance()
            
            const teacher = mockResponseData({user:{id: '1', first_name: 'teacher name', username: 'email@gmail.com'}, teacher:{id:'12'}})
            axios.get.mockImplementationOnce(() =>  Promise.resolve(teacher))

            instance.getTeacherClasses = jest.fn()
            instance.getTeacherClasses.mockReturnValueOnce()


            await instance.componentDidMount()

            expect(instance.state.first_name).toBe('teacher name')
            expect(instance.getTeacherClasses).toHaveBeenCalledTimes(1)
            
        })
    })

    describe('Profile Classes', ()=>{
        it('should get teacher classes', async()=>{

            const instance = wrapper.instance()
            const classrooms = mockResponseData('classrooms')
            axios.get.mockImplementationOnce(() =>  Promise.resolve(classrooms))

            instance.findTeacherClassrooms = jest.fn()
            instance.findTeacherClassrooms.mockReturnValueOnce()
            await instance.getTeacherClasses()
            expect(instance.findTeacherClassrooms).toHaveBeenCalledTimes(1)
        })

        it('should find teacher classes', ()=>{
            const instance = wrapper.instance()
            const classrooms = [{teacher_id: '2', name: 'name'}, {teacher_id: '3', name: 'name2'}]
            instance.state.teacher_id = '2'
            instance.findTeacherClassrooms(classrooms)
            expect(instance.state.classes).toStrictEqual([{teacher_id: '2', name: 'name'}])
        })

        it('should not find a teacher classe', ()=>{
            const instance = wrapper.instance()
            const classrooms = [{teacher_id: '5', name: 'name'}, {teacher_id: '3', name: 'name2'}]
            instance.state.teacher_id = '2'
            instance.findTeacherClassrooms(classrooms)
            expect(instance.state.classes).toStrictEqual([])
        })
    })

    describe('Student Information', ()=>{
        it('should get student balance', async()=>{
            const instance = wrapper.instance()
            
            const balance = mockResponseData('balance')
            axios.get.mockImplementationOnce(() =>  Promise.resolve(balance))

            await instance.getStudentBalance()
            expect(instance.state.balance).toBe('balance')
        })

        it('should get student transactions', async()=>{
            const instance = wrapper.instance()
            
            const transactions = mockResponseData(['transactions', 'transactions2'])
            axios.get.mockImplementationOnce(() =>  Promise.resolve(transactions))

            await instance.getStudentTransactions()
            expect(instance.state.transactions).toStrictEqual(['transactions', 'transactions2'])
        })

        it('should get student bought items', async()=>{
            const instance = wrapper.instance()
            
            const items = mockResponseData(['item', 'item2'])
            axios.get.mockImplementationOnce(() =>  Promise.resolve(items))

            await instance.getBoughtItems()
            expect(instance.state.bought_items).toStrictEqual(['item', 'item2'])
        })
    })

    describe('Class Creation', ()=>{
        it('Should create a class', async()=>{
            const instance = wrapper.instance()
            const classrooms = mockResponseData('classrooms')
            axios.get.mockImplementationOnce(() =>  Promise.resolve(classrooms))
            const classreturn = {class: '1'}
            axios.post.mockImplementationOnce(() =>  Promise.resolve(classreturn))

            instance.state.new_class_name = 'new class'
            instance.state.teacher_id = '2'
            
            instance.makeClassCode = jest.fn()
            instance.makeClassCode.mockReturnValueOnce('ABCDEF')

            await instance.createClass()
            
            expect(instance.state.showCreateClass).toBe(false)
            const payload = {class_name: 'new class', teacher_id: '2', class_code: 'ABCDEF'}
            expect(axios.get).toHaveBeenCalledTimes(1)
            expect(axios.post).toHaveBeenCalledTimes(1)
            expect(axios.post).toHaveBeenCalledWith("https://mythical-mason-324813.ey.r.appspot.com//classrooms/", payload)
            expect(instance.state.new_class_code).toBe('ABCDEF')
        })

        it('should make a new class code', ()=>{
            const instance = wrapper.instance()

            const classrooms = [{class_code: '123456'}, {class_code: 'BBBBBB'}]
            const result = instance.makeClassCode(classrooms)
            expect(result.length).toBe(6)
        })
    })

    describe('Page Renders', ()=>{
        it('should render the student profile correctly', ()=>{
            expect(wrapper.getElements()).toMatchSnapshot()
        })

        it('should render the teacher profile correctly', ()=>{
            expect(wrapperTeacher.getElements()).toMatchSnapshot()
        })

        it('should render the inventory modal correctly', ()=>{
            const instance = wrapper.instance()
            instance.state.showInven = true
            instance.state.bought_items = [{id: '1', item_name: 'name'}, {id: "2", item_name: 'name2'}]
            expect(instance.renderInventoryModal()).toMatchSnapshot()
        })

        it('should render card that is a positive', ()=>{
            const instance = wrapper.instance()
            const item = {symbol: '+', name: 'name', amount: '100'}
            expect(instance.renderCard(item)).toBe('From name +100$')
        })

        it('should render card that is a negative', ()=>{
            const instance = wrapper.instance()
            const item = {symbol: '-', name: 'name', amount: '100'}
            expect(instance.renderCard(item)).toBe('To name -100$')
        })
        
        
    })

})

const mockResponseData = jest.fn((data) => {
    if (typeof data === 'string')
    {
        return {
            data: data,
        };
    }
    else
    {
        return {
            data: data,
        };
    }
  });