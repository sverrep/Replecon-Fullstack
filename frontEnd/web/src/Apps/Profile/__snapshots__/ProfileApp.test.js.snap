// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Student Profile Tests Page Renders should render the inventory modal correctly 1`] = `
<Modal
  animation={true}
  autoFocus={true}
  backdrop={true}
  dialogAs={
    Object {
      "$$typeof": Symbol(react.forward_ref),
      "render": [Function],
    }
  }
  enforceFocus={true}
  keyboard={false}
  restoreFocus={true}
  show={true}
>
  <ModalHeader
    closeButton={false}
    closeLabel="Close"
  >
    <ModalTitle>
      My Inventory
    </ModalTitle>
    <Button
      active={false}
      disabled={false}
      onClick={[Function]}
      variant="secondary"
    >
      Close
    </Button>
  </ModalHeader>
  <ModalBody>
    <Form>
      <div
        className="boughtItems"
      >
        <ul>
          <li
            className="grey cardb"
          >
            name
          </li>
          <li
            className="grey cardb"
          >
            name2
          </li>
        </ul>
      </div>
    </Form>
  </ModalBody>
</Modal>
`;

exports[`Student Profile Tests Page Renders should render the student profile correctly 1`] = `
Array [
  <div
    className="wrapper"
  >
    <Navbar
      className="teacher-navbar"
      collapseOnSelect={true}
      expand="lg"
      variant="dark"
    >
      <Container
        fluid={false}
      >
        <NavbarBrand
          className="navbar-header"
        >
          Replecon
        </NavbarBrand>
        <Nav
          className="navbar-btns"
          fill={false}
          justify={false}
        >
          <Button
            active={false}
            className="navbar-btn"
            disabled={false}
            variant="light"
          >
            <Link
              className="navbar-link"
              to={
                Object {
                  "pathname": "/Profile",
                  "state": Object {
                    "role": "Student",
                    "token": "Token",
                  },
                }
              }
            >
              Profile
            </Link>
          </Button>
          <Button
            active={false}
            className="navbar-btn"
            disabled={false}
            variant="light"
          >
            <Link
              className="navbar-link"
              to={
                Object {
                  "pathname": "/StudentStore",
                  "state": Object {
                    "role": "Student",
                    "token": "Token",
                  },
                }
              }
            >
              Store
            </Link>
          </Button>
          <Button
            active={false}
            className="navbar-btn"
            disabled={false}
            variant="light"
          >
            <Link
              className="navbar-link"
              to={
                Object {
                  "pathname": "/StudentBank",
                  "state": Object {
                    "role": "Student",
                    "token": "Token",
                  },
                }
              }
            >
              Bank
            </Link>
          </Button>
          <Button
            active={false}
            className="navbar-btn"
            disabled={false}
            variant="light"
          >
            <Link
              className="navbar-link"
              to={
                Object {
                  "pathname": "/StudentClass",
                  "state": Object {
                    "role": "Student",
                    "token": "Token",
                  },
                }
              }
            >
              Classroom
            </Link>
          </Button>
        </Nav>
      </Container>
    </Navbar>
    <div
      className="profile-content"
    >
      <div
        className="profile-title"
      >
        <div
          className="profile-name"
        >
          <h3>
            
          </h3>
        </div>
        <div
          className="profile-balance"
        >
          <Button
            active={false}
            className="profile-button"
            disabled={false}
            onClick={[Function]}
            variant="primary"
          >
            Log Out
          </Button>
          <Button
            active={false}
            className="inventory-button"
            disabled={false}
            onClick={[Function]}
            variant="primary"
          >
            Inventory
          </Button>
        </div>
        <Modal
          animation={true}
          autoFocus={true}
          backdrop={true}
          dialogAs={
            Object {
              "$$typeof": Symbol(react.forward_ref),
              "render": [Function],
            }
          }
          enforceFocus={true}
          keyboard={false}
          restoreFocus={true}
          show={false}
        >
          <ModalHeader
            closeButton={false}
            closeLabel="Close"
          >
            <ModalTitle>
              My Inventory
            </ModalTitle>
            <Button
              active={false}
              disabled={false}
              onClick={[Function]}
              variant="secondary"
            >
              Close
            </Button>
          </ModalHeader>
          <ModalBody>
            <Form>
              <div
                className="boughtItems"
              >
                <ul />
              </div>
            </Form>
          </ModalBody>
        </Modal>
      </div>
      <h3>
        $
        
      </h3>
      <div
        className="content"
      >
        <div>
          <ul />
        </div>
      </div>
    </div>
  </div>,
]
`;

exports[`Student Profile Tests Page Renders should render the teacher profile correctly 1`] = `
Array [
  <Container
    className="teacher-container"
    fluid={false}
  >
    <Row>
      <h3
        className="teacher-h3"
      >
        Welcome Back 
        
      </h3>
      <h4>
        Your Classes
      </h4>
      <div
        className="teacher-btns-row"
      >
        <Button
          active={false}
          className="teacher-btn-create"
          disabled={false}
          onClick={[Function]}
          variant="primary"
        >
          Create New Class
        </Button>
      </div>
      <div
        className="teacher-btns-row"
      >
        <Button
          active={false}
          className="teacher-btn-logout"
          disabled={false}
          onClick={[Function]}
          variant="outline-secondary"
        >
          Log Out
        </Button>
      </div>
    </Row>
    <Modal
      animation={true}
      autoFocus={true}
      backdrop={true}
      centered={true}
      dialogAs={
        Object {
          "$$typeof": Symbol(react.forward_ref),
          "render": [Function],
        }
      }
      enforceFocus={true}
      keyboard={true}
      onHide={[Function]}
      restoreFocus={true}
      show={false}
      size="lg"
    >
      <ModalHeader
        closeButton={true}
        closeLabel="Close"
      >
        <ModalTitle>
          Create A Class
        </ModalTitle>
      </ModalHeader>
      <ModalBody>
        <FloatingLabel
          label="Class Name"
        >
          <FormControl
            id="new_class_name"
            onChange={[Function]}
            placeholder="Class Name"
          />
        </FloatingLabel>
      </ModalBody>
      <ModalFooter>
        <Button
          active={false}
          className="modal-btn"
          disabled={false}
          onClick={[Function]}
          variant="primary"
        >
          Create Class
        </Button>
      </ModalFooter>
    </Modal>
  </Container>,
]
`;
